### PR TITLE
[Test] Add multi-fabric test for 15831

### PR DESCRIPTION
**Pull request by erjiaqing**
#### Problem
Tests in #15831 does not contains the test cases for real multi fabric operations in unit tests.

#### Change overview
Adds the test for multi fabric operations

#### Testing
- Only updates tests and some helper functions.

### CODE REVIEW

**Review by andy31415 — APPROVED** [review]

**Review by tcarmelveilleux — APPROVED** [review]

File: src/app/InteractionModelEngine.h [hunk 1]
```diff
@@ -137,6 +137,7 @@ class InteractionModelEngine : public Messaging::UnsolicitedMessageHandler,
 
     uint32_t GetNumActiveReadHandlers() const;
     uint32_t GetNumActiveReadHandlers(ReadHandler::InteractionType type) const;
+    uint32_t GetNumActiveReadHandlers(ReadHandler::InteractionType type, FabricIndex fabricIndex) const;
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.h:140]: Please document.

File: src/controller/tests/data_model/TestRead.cpp [hunk 2]
```diff
@@ -1617,12 +1615,76 @@ void TestReadInteraction::TestReadHandler_KillOverQuotaSubscriptions(nlTestSuite
         app::InteractionModelEngine::GetInstance()->GetReportingEngine().SetDirty(path);
     }
     readCallback.ClearCounters();
-    ctx.DrainAndServiceIO();
 
-    NL_TEST_ASSERT(apSuite, readCallback.mAttributeCount <= kExpectedParallelPaths);
+    // Run until the global dirtyset is cleared.
```
Comment (bzbarsky-apple) [inline src/controller/tests/data_model/TestRead.cpp:1619]: Why does `DrainAndServiceIO` not handle this correctly?  I thought it handled ScheduleWork called from inside something else it processes....

File: src/controller/tests/data_model/TestRead.cpp [hunk 3]
```diff
@@ -1617,12 +1615,76 @@ void TestReadInteraction::TestReadHandler_KillOverQuotaSubscriptions(nlTestSuite
...
+    ctx.GetIOContext().DriveIOUntil(System::Clock::Seconds16(60), [&]() {
+        return app::InteractionModelEngine::GetInstance()->GetReportingEngine().GetGlobalDirtySetSize() == 0 &&
+            app::InteractionModelEngine::GetInstance()->GetReportingEngine().GetNumReportsInFlight() == 0;
+    });
+
+    // We should evict the subscription with exceeded resources, so we should used exactly all resources.
```
Comment (bzbarsky-apple) [inline src/controller/tests/data_model/TestRead.cpp:1625]:
```suggestion
    // We should evict the subscriptions with excess resources, so we should use exactly all resources.
```

File: src/controller/tests/data_model/TestRead.cpp [hunk 4]
```diff
@@ -1617,12 +1615,76 @@ void TestReadInteraction::TestReadHandler_KillOverQuotaSubscriptions(nlTestSuite
...
-                       static_cast<size_t>(kExpectedParallelSubs));
+                   app::InteractionModelEngine::GetInstance()->GetNumActiveReadHandlers() ==
+                       static_cast<uint32_t>(kExpectedParallelSubs));
+
+    NL_TEST_ASSERT(apSuite,
+                   app::InteractionModelEngine::GetInstance()->GetNumActiveReadHandlers(
+                       app::ReadHandler::InteractionType::Subscribe, ctx.GetAliceFabricIndex()) >
+                       app::InteractionModelEngine::kMinSupportedSubscriptionsPerFabric);
+
+    // The following check will trigger the logic in im to kill the read handlers that uses more paths than the limit per fabric.
```
Comment (bzbarsky-apple) [inline src/controller/tests/data_model/TestRead.cpp:1636]:
```suggestion
    // The following check will trigger the logic in im to kill the read handlers that use more paths than the limit per fabric.
```

**Review by bzbarsky-apple — APPROVED** [review]

**Review by yunhanw-google — APPROVED** [review]

**Review by emargolis — APPROVED** [review]

File: src/controller/tests/data_model/TestRead.cpp (hunk 2 above)
Comment (erjiaqing) [reply src/controller/tests/data_model/TestRead.cpp:1619]: DrainAndServiceIO requires "network traffic" between "nodes", if some function called "ScheduleWork" then DrainAndServiceIO will see no traffic and return.

File: src/app/InteractionModelEngine.h (hunk 1 above)
Comment (erjiaqing) [reply src/app/InteractionModelEngine.h:144]: Added

**Pull request merged**